src/Dictionaries/CacheDictionaryUpdateQueue.cpp

```diff
@@ -35,40 +36,56 @@ CacheDictionaryUpdateQueue<dictionary_key_type>::CacheDictionaryUpdateQueue(
 template <DictionaryKeyType dictionary_key_type>
 CacheDictionaryUpdateQueue<dictionary_key_type>::~CacheDictionaryUpdateQueue()
 {
-    try {
-        if (!finished)
-            stopAndWait();
+    if (finished)
+        return;
+
+    try
+    {
+        stopAndWait();
     }
     catch (...)
     {
-        /// TODO: Write log
+        tryLogCurrentException(log, "Caught exception during destruction");
     }
 }
 
 template <DictionaryKeyType dictionary_key_type>
 void CacheDictionaryUpdateQueue<dictionary_key_type>::tryPushToUpdateQueueOrThrow(CacheDictionaryUpdateUnitPtr<dictionary_key_type> & update_unit_ptr)
 {
     if (finished)
-        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue finished");
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue for dictionary {} already finished", dictionary_name_for_logs);
 
-    if (!update_queue.tryPush(update_unit_ptr, configuration.update_queue_push_timeout_milliseconds))
+    if (!empty_count.tryWait(configuration.update_queue_push_timeout_milliseconds))
         throw DB::Exception(ErrorCodes::CACHE_DICTIONARY_UPDATE_FAIL,
             "Cannot push to internal update queue in dictionary {}. "
             "Timelimit of {} ms. exceeded. Current queue size is {}",
             dictionary_name_for_logs,
-            std::to_string(configuration.update_queue_push_timeout_milliseconds),
-            std::to_string(update_queue.size()));
+            configuration.update_queue_push_timeout_milliseconds,
+            getSize());
+
+    try
+    {
+        std::unique_lock<std::mutex> update_lock(queue_mutex);
+        queue.push(update_unit_ptr);
+
+        queue_cond.notify_one();
+    }
+    catch (...)
+    {
+        empty_count.set();
+        throw;
+    }
 }
 
 template <DictionaryKeyType dictionary_key_type>
 void CacheDictionaryUpdateQueue<dictionary_key_type>::waitForCurrentUpdateFinish(CacheDictionaryUpdateUnitPtr<dictionary_key_type> & update_unit_ptr) const
 {
     if (finished)
-        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue finished");
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "CacheDictionaryUpdateQueue for dictionary {} already finished", dictionary_name_for_logs);
```

Comment:
Why `UNSUPPORTED_METHOD`, looks like `LOGICAL_ERROR` is better?

Reply:
Fixed.